sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroSink.java

```diff
@@ -115,17 +128,35 @@ protected void prepareWrite(WritableByteChannel channel) throws Exception {
                   + v.getClass().getSimpleName());
         }
       }
-      dataFileWriter.create(schema, Channels.newOutputStream(channel));
+      if (schema != null){ //else lazy init
+        dataFileWriter.create(schema, Channels.newOutputStream(channel));
+      }
     }
 
     @Override
     public void write(OutputT value) throws Exception {
+      // lazy init dataFileWriter at first write once we have the value to get the schema from
```

Comment:
I very strongly disagree with this approach. If the user applies this to a PCollection that has records with different schemas (which is not at all difficult to do by accident - just do two readGenericRecords() with different schema and Flatten them), this will non-deterministically either fail (if records with different schema end up in the same bundle, I suppose DataFileWriter verifies compliance of each new record with the current schema?), or silently produce garbage (if records with different schema end up in different bundles, you'll generate some Avro files with one schema and some with a different schema).

I realize that in your use case you were able to structure your pipelines in a way that this doesn't happen, but future users of this API might use it in unanticipated ways, and I don't want to add an API that has so few safeguards against silently producing garbage. Records with potentially different schemas *must* be grouped by schema before writing and I think DynamicDestinations is the already existing and proper way to do this - if you feel that it is unnecessarily verbose, let's discuss alternative ways of addressing that verbosity.

Reply:
I understand your points and I agree. 

As I wrote in the description of this PR, I know that this implementation requires all the elements in the PCollection to share the same schema. What I don't understand is what I asked in the description of the PR and in the ticket: 

How is the current `AvroIO.write(schema)` implementation different?
   
If a user uses current `AvroIO.write(schema)` implementation, then this schema instance is passed to the `TypedWrite` then to the `ConstantAvroDestination` that is used in `AvroSink` which does the write. So all the elements need to share the same schema here too. So the current implementation should behave the same as this PR if the user provides a PCollection with heterogeneous schemas (and different from the one specified) no?
=> Please tell me what I am missing.

For the other point: indeed you're right, this PR is mainly to avoid verbosity :) because as I said before, using `DynamicDestinations` and a schema side input works perfectly!

Reply:
The difference between AvroIO.write(schema) and the current PR is that if you pass a collection with different schemas (or with the wrong schema) to AvroIO.write(schema), then the pipeline will always fail and no garbage data will be written. If there's a way to provide the same guarantee with the current approach, I'd be more open to it, but I don't yet see how it can be done.

In terms of avoiding verbosity: do you think that would be addressed by FileIO.write() better than current AvroIO.write()?

See https://github.com/apache/beam/pull/3817 . Writing to Avro with that PR can look like:

PCollection<GenericRecord> records = ...;

records.apply(FileIO.write().to(...).via(fn(
    c -> AvroIO.sinkViaGenericRecords(c.sideInput(schemaView)), 
    requiresSideInputs(schemaView))));

Reply:
- Indeed there is a particular case in which the PR behaves differently: as the PCollection is not ordered, if one bundle ends up having only SCHEMA1 records and the other only SCHEMA2 records, then the implementation in this PR will write the 2 bundles with no error because it will guess SCHEMA1 from bundle 1 and SCHEMA2 from bundle 2 that will probably result in producing an avro file that has 2 schemas.
**How is the concatenation of temporary bundle avro files done? Is there a schema check there? Maybe we could provide the guarantee by having a check at concatenation time to avoid producing an avro file that will contain 2 schemas in that particular case. And after it, once secured, adding to the javadoc that it only supports one schema per collection will be enough. WDYT?**
- in the other cases (bundles with mixed schemas): the current implementation and this PR will fail with the same avro exception when encoding the first record in the bundle that has a different schema than specified/guessed (see the tests added in last commits). 
- #3817 and the `FileIO` new architecture is indeed an improvement to the verbosity among other improvements.

Reply:
Let's discuss the point above before I rebase on master (AvroTransformTest deleted)

Reply:
@jkff WDYT?

Reply:
Temporary files produced by bundles are not concatenated - there is at least 1 file per bundle. I understand the temptation to lazily infer the schema from a "first" element, but I don't believe that it's possible to implement safely. This is something that is fine to do for a user in their own pipeline (or in their own library), but it's not a level of safety acceptable in the Beam SDK itself.

I generally like the idea of an AvroIO.write() where you don't need to know the schema of the whole write() at construction time e.g. because the input PCollection may contain different schemas - but then you need to be explicit about 1) partitioning the records into subsets that have the same schema, and 2) somehow enforcing that the schema within a subset is in fact the same. I believe this is exactly what DynamicDestinations already allows you to do, and FileIO.write() will allow to do with a lot fewer boilerplate.

So, briefly speaking, I am against the direction in which this PR is heading and I don't think it's fixable :-| 

WDYT about the following:

- For now, use DynamicDestinations and later FileIO.write() in your code
- Some time later, let's revisit the issues you faced while doing that, and try to figure out a better way to solve them

?

Reply:
Ok fair enough. Thank you for your opinion. I'm closing this PR. I had no issues in particular, just tried to avoid having the shcema at construction time but it can be done using `DynamicDestinations` as I did there https://github.com/echauchot/beam/blob/671e74b7934050c31e8fb46f1e9ced9da9ca867e/sdks/java/core/src/test/java/org/apache/beam/sdk/io/AvroIOTransformTest.java#L393 . It was mainly a matter of verbosity like we discussed. 
I'll rewrite the above axample using `FileIO.write()` when https://github.com/apache/beam/pull/3817 is merged.